In [15]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.impute import KNNImputer
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import SelectKBest, f_regression

In [2]:
df=pd.read_csv('../data/Automobile_data.csv')

# Mostrar las primeras filas del dataset
print("Primeras filas del dataset:")
df.head()

Primeras filas del dataset:


symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [3]:
# Información general del dataset
print("\nInformación del dataset:")
df.info()


Información del dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  

In [4]:
# Estadísticas descriptivas iniciales
print("\nEstadísticas descriptivas iniciales:")
df.describe()


Estadísticas descriptivas iniciales:


symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

In [5]:
# Identificar duplicados
duplicados = df.duplicated()
# Contar el número de duplicados
num_duplicados = duplicados.sum()
print(f"Número de registros duplicados: {num_duplicados}")
df.head()

Número de registros duplicados: 0


symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [6]:
# Verificacion de tipo de datos
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

In [7]:
columnas_categoricas = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels',
                        'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders',
                        'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg',
                        'price']
for columna in columnas_categoricas:
    print(f"Valores únicos en '{columna}':")
    print(df[columna].unique())
    print("\n")

Valores únicos en 'symboling':
[ 3  1  2  0 -1 -2]


Valores únicos en 'normalized-losses':
['?' '164' '158' '192' '188' '121' '98' '81' '118' '148' '110' '145' '137'
 '101' '78' '106' '85' '107' '104' '113' '150' '129' '115' '93' '142'
 '161' '153' '125' '128' '122' '103' '168' '108' '194' '231' '119' '154'
 '74' '186' '83' '102' '89' '87' '77' '91' '134' '65' '197' '90' '94'
 '256' '95']


Valores únicos en 'make':
['alfa-romero' 'audi' 'bmw' 'chevrolet' 'dodge' 'honda' 'isuzu' 'jaguar'
 'mazda' 'mercedes-benz' 'mercury' 'mitsubishi' 'nissan' 'peugot'
 'plymouth' 'porsche' 'renault' 'saab' 'subaru' 'toyota' 'volkswagen'
 'volvo']


Valores únicos en 'fuel-type':
['gas' 'diesel']


Valores únicos en 'aspiration':
['std' 'turbo']


Valores únicos en 'num-of-doors':
['two' 'four' '?']


Valores únicos en 'body-style':
['convertible' 'hatchback' 'sedan' 'wagon' 'hardtop']


Valores únicos en 'drive-wheels':
['rwd' 'fwd' '4wd']


Valores únicos en 'engine-location':
['front' 'rear']


Val

In [8]:
# Identificar valores faltantes
df.isna().sum()

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64

In [11]:
# Reemplazar '?' en la columna 'price' con la mediana de los valores numéricos
df['price'] = df['price'].replace('?', None).astype(float)
price_median = df['price'].median()
df['price'] = df['price'].fillna(price_median)

# Confirmar que no existen valores no válidos en 'price'
print("¿Valores no numéricos en 'price'?", df['price'].isnull().sum())

¿Valores no numéricos en 'price'? 0


In [12]:


# Reemplazar '?' por un valor válido (usaremos la mediana de la columna)
for column in df.columns:
    if df[column].dtype == 'object':  # Si es categórica
        df[column] = df[column].replace('?', df[column].mode()[0])  # Modo para categóricas
    else:  # Si es numérica
        df[column] = pd.to_numeric(df[column], errors='coerce')  # Convertir a numérico
        df[column] = df[column].fillna(df[column].median())  # Mediana para numéricas

# Codificar variables categóricas
df_encoded = pd.get_dummies(df, drop_first=True)

# Dividir en X (características) e y (objetivo)
X = df_encoded.drop('price', axis=1)
y = df_encoded['price']

# División entrenamiento/prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Escalar las características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Modelos
models = {
    'Linear Regression': LinearRegression(),
    'KNN': KNeighborsRegressor(n_neighbors=5),
    'Decision Tree': DecisionTreeRegressor(random_state=42)
}

# Entrenar y evaluar
results = {}
for model_name, model in models.items():
    model.fit(X_train_scaled, y_train)
    y_pred = model.predict(X_test_scaled)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    results[model_name] = {'MSE': mse, 'R2': r2}

# Mostrar resultados
for model_name, metrics in results.items():
    print(f"{model_name} - MSE: {metrics['MSE']:.2f}, R²: {metrics['R2']:.2f}")


Linear Regression - MSE: 756473379034200356702749018030080.00, R²: -9643555409022041842515968.00
KNN - MSE: 36736475.00, R²: 0.53
Decision Tree - MSE: 7481248.88, R²: 0.90


In [ ]:

# Identificar las características más importantes
selector = SelectKBest(score_func=f_regression, k=10)  # Seleccionar las 10 mejores características
X_new = selector.fit_transform(X, y)

# Revisar las características seleccionadas
selected_features = X.columns[selector.get_support()]
print("Características seleccionadas:", selected_features)

# Dividir de nuevo los datos
X_train, X_test, y_train, y_test = train_test_split(X_new, y, test_size=0.2, random_state=42)

# Entrenar el modelo con las nuevas características
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

model_lr = LinearRegression()
model_lr.fit(X_train_scaled, y_train)

# Predicciones y evaluación
y_pred_lr = model_lr.predict(X_test_scaled)
mse_lr = mean_squared_error(y_test, y_pred_lr)
r2_lr = r2_score(y_test, y_pred_lr)

print(f"Regresión Lineal Mejorada - MSE: {mse_lr}, R²: {r2_lr}")
print(df.columns)


Características seleccionadas: Index(['wheel-base', 'length', 'width', 'curb-weight', 'engine-size',
       'city-mpg', 'highway-mpg', 'drive-wheels_fwd', 'drive-wheels_rwd',
       'num-of-cylinders_four'],
      dtype='object')
Regresión Lineal Mejorada - MSE: 13092610.04909001, R²: 0.8330948398761695
Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


### Discusión sobre las diferencias en el rendimiento:

Regresión Lineal Mejorada: Después de la optimización, la regresión lineal muestra un MSE más bajo (13,092,610) y un R² relativamente alto (0.83). Este modelo es bastante adecuado, pero es sensible a las relaciones lineales en los datos.

Regresión Lineal (Primer intento): El primer intento de regresión lineal dio resultados extremos, con un MSE extremadamente alto y un R² negativo, lo que indica un mal ajuste del modelo a los datos.

KNN (K-Nearest Neighbors): Aunque KNN tiene un MSE razonable (36,736,475), su R² es mucho más bajo (0.53), lo que indica que el modelo no está capturando muy bien la variabilidad de los precios.

Árbol de Decisión: Este modelo tiene el MSE más bajo (7,481,248) y el R² más alto (0.90), lo que sugiere que el Árbol de Decisión es el mejor modelo en términos de precisión y capacidad de ajuste a los datos.

Interpretación de Resultados

Modelo más adecuado para el problema de predicción de precios de autos:

Árbol de Decisión es el modelo que mejor se ajusta a los datos, con el MSE más bajo y el R² más alto. Esto sugiere que el modelo es capaz de capturar las relaciones no lineales entre las características del auto y el precio, lo que lo hace adecuado para este tipo de predicción.
Posibles mejoras y próximos pasos:

Ajuste de Hiperparámetros: Para mejorar aún más el rendimiento de los modelos, se pueden ajustar los hiperparámetros. Esto es particularmente útil en modelos como KNN y Árboles de Decisión, que dependen de parámetros como el número de vecinos (K) y la profundidad del árbol.

Uso de Técnicas Avanzadas: Técnicas más avanzadas como Redes Neuronales o Modelos de Ensamblaje (por ejemplo, Random Forest o Gradient Boosting), que combinan varios modelos, pueden mejorar aún más la predicción al aprovechar las fortalezas de cada uno.

Ingeniería de Características: Se puede realizar una mayor exploración y transformación de las características, como la creación de variables interactivas o la normalización de variables numéricas, lo que podría mejorar el rendimiento de los modelos.

Validación Cruzada: Implementar validación cruzada para evaluar la estabilidad del modelo y evitar el sobreajuste (overfitting).